### DISCUSSION TITLE
unpack解析错误

### DISCUSSION THREAD

**Discussion by novada0** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#69] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/107106546/211450338-e8569ae0-6143-4cd3-99bd-4853b5b4e0f7.png)
![image](https://user-images.githubusercontent.com/107106546/211450381-fc327849-b6a1-4bb2-bfae-f0fabf2516b3.png)
![image](https://user-images.githubusercontent.com/107106546/211450423-3c3539f4-86e0-4be4-95a2-c623bd5e4938.png)
错误截图：
![image](https://user-images.githubusercontent.com/107106546/211450626-0d242cd6-e5e7-4d39-9ed4-0b938b518430.png)
学长，求助一个问题。我把时间戳存在header里面，发出去的包没问题，但是接收到的包说[字节不匹配](struct.error: unpack requires a buffer of 28 bytes)，明明应该是32个字节，为啥这里就解析成28个字节了？我用的系统是[Ubuntu 20.04]，python3.8.10。

**Comment by novada0** [comment]
这是发出去包的header：
![image](https://user-images.githubusercontent.com/107106546/211451110-93a14565-7cbd-437f-a0dd-8df88547d585.png)

**Reply by TANSixu** [reply]
Hi，请问HEADER_LEN有更新成现在的长度嘛

看起来长度应该的确是28：2(H)+1(B)+1(B)+2(H)+2(H)+4(I)+4(I)+4(I)+8(Q) = 2+1+1+2+2+4+4+4+8 = 28

**Comment by novada0** [comment]
![image](https://user-images.githubusercontent.com/107106546/211480100-ea0c40b5-28d8-4a07-bdbb-8f41ca15e30c.png)
![image](https://user-images.githubusercontent.com/107106546/211480132-cd0b381e-ba57-4213-8a2a-fde39e108bbd.png)
为什么这样会算出来32呢

**Comment by novada0** [comment]
噢噢，刚刚试了一下。
![image](https://user-images.githubusercontent.com/107106546/211480355-d2d1eed4-d597-4467-8a82-543501977ecd.png)
这样子的输出才是正确的。谢谢学长！

**Comment by TANSixu** [comment, marked as answer]
突然发现这是个非常好的问题！之前我们忘记了提示，就是pack会自动进行代码对齐。文档中提到：
<img width="490" alt="image" src="https://user-images.githubusercontent.com/66773369/211529499-7533f644-8535-4093-a974-4f2c0c13abcc.png">
且如果没有在fmt字符串中加入任何特殊首字符，会默认为@：
```
如果第一个字符不是其中之一，则假定为 '@' 。
```
所以这个fmt在calcsize中会得到32，就是因为自动对齐了。指定！为首字符后，不会进行对齐。如果有修改header的同学可以注意一下这个地方：自动对齐后的变量位置或者如何手动对齐。
